File: trunk/FireflowSuite/engine/src/org/fireflow/engine/RuntimeContext.java
package org.fireflow.engine;

import java.util.HashMap;
import java.util.Map;

import java.util.logging.Level;
import java.util.logging.Logger;
import org.fireflow.engine.calendar.ICalendarService;
import org.fireflow.engine.definition.IDefinitionService;
import org.fireflow.engine.impl.WorkflowSession;
import org.fireflow.engine.persistence.IPersistenceService;
import org.fireflow.engine.taskinstance.ITaskInstanceManager;
import org.fireflow.kenel.KenelException;
import org.fireflow.kenel.KenelManager;
import org.fireflow.engine.condition.IConditionResolver;


/**
 * RuntimeContext提供所有流程实例运行环境
 * @author chennieyun
 *
 */
public class RuntimeContext {

    private static RuntimeContext instance = null;
    private boolean isInitialized = false;
    
    //context管理的各种服务
    private IConditionResolver conditionResolver = null;
    private IPersistenceService persistenceService = null;
    private IDefinitionService definitionService = null;
    private KenelManager kenelManager = null;
    private ITaskInstanceManager taskInstanceManager = null;
    private ICalendarService calendarService = null;
    
    private ThreadLocal<Object> currentDBSession = new ThreadLocal<Object>();
//	private ThreadLocal<IFireflowSession> currentFireflowSession = new ThreadLocal<IFireflowSession>();
    
    //TODO
    //保持一个内部用的beanFactory,类似spring的beanfactory,用于注册Listener,ApplicationHandler,AssignmentHandler的
    //单态实例，可以考虑这个beanfactory是否可以引用spring的beanfactory
    private Map<String, Object> beanCache = new HashMap<String, Object>();

    //为了使kenel能够用于IDE环境，runtimecontext不要采用此模式，而是通过token区获取。
//    public static RuntimeContext getInstance() {
//        if (instance == null) {
//            instance = new RuntimeContext();
//        }
//        return instance;
//    }


    public RuntimeContext() {
    }

    /**
     * 该方法用户缓存诸如事件监听器，applicationHandler实例等。
     * 不过，这种实现不是很漂亮。
     * @param beanClassName
     * @return
     */
    public Object getBeanByClassName(String beanClassName) {
        Object obj = beanCache.get(beanClassName);
        if (obj != null) {
            return obj;
        }
        Class clz;
        try {
            clz = Class.forName(beanClassName);
            obj = clz.newInstance();
        } catch (ClassNotFoundException ex) {
            Logger.getLogger(RuntimeContext.class.getName()).log(Level.SEVERE, null, ex);
        } catch (InstantiationException ex) {
            Logger.getLogger(RuntimeContext.class.getName()).log(Level.SEVERE, null, ex);
        } catch (IllegalAccessException ex) {
            Logger.getLogger(RuntimeContext.class.getName()).log(Level.SEVERE, null, ex);
        }
        beanCache.put(beanClassName, obj);
        return obj;
    }

    //DBSession 放在哪里比较好些？
//    public void setCurrentDBSession(Object dbSession) {
//        currentDBSession.set(dbSession);
//    }
//
//    public Object getCurrentDBSession() {
//        return currentDBSession.get();
//    }
    
    public IConditionResolver getConditionResolver() {
        return conditionResolver;
    }

    public void setConditionResolver(IConditionResolver resolver) {
        this.conditionResolver = resolver;
        if (this.conditionResolver instanceof IRuntimeContextAware){
            ((IRuntimeContextAware)this.conditionResolver).setRuntimeContext(this);
        }
    }

    public ITaskInstanceManager getTaskInstanceManager() {
        return taskInstanceManager;
    }

    public void setTaskInstanceManager(ITaskInstanceManager taskInstanceManager) {
        this.taskInstanceManager = taskInstanceManager;
        this.taskInstanceManager.setRuntimeContext(this);
    }

    public IPersistenceService getPersistenceService() {
        return this.persistenceService;
    }

    public void setPersistenceService(IPersistenceService persistenceService) {
        this.persistenceService = persistenceService;
        this.persistenceService.setRuntimeContext(this);
    }

    public IDefinitionService getDefinitionService() {
        return definitionService;
    }

    public void setDefinitionService(IDefinitionService service) {
        this.definitionService = service;
        this.definitionService.setRuntimeContext(this);
    }

    public KenelManager getKenelManager() {
        return kenelManager;
    }

    public void setKenelManager(KenelManager kenelManager) {
        this.kenelManager = kenelManager;
        this.kenelManager.setRuntimeContext(this);
    }

    public ICalendarService getCalendarService() {
        return calendarService;
    }

    public void setCalendarService(ICalendarService calendarService) {
        this.calendarService = calendarService;
        this.calendarService.setRuntimeContext(this);
    }

    

    public void initialize() throws EngineException, KenelException {
//		System.out.println("执行initialize(),the isInitialized="+this.isInitialized);
//		System.out.println("看看有没有 受理环节实例"+this.getKenelManager().getWFElementInstance("受理"));
        if (!isInitialized) {
//			initKenelExtensions();
            initAllNetInstances();
            isInitialized = true;
        }
    }

 
    protected void initAllNetInstances() throws KenelException {
        /*
        List<WorkflowProcess> allWfProcess = definitionService.getAllWorkflowProcesses();
        this.getKenelManager().clearAllNetInstance();
        for(int i=0;allWfProcess!=null && i<allWfProcess.size();i++){
        WorkflowProcess process = allWfProcess.get(i);
        this.getKenelManager().createNetInstance( process);
        }
         */
    }
    
    public IWorkflowSession getWorkflowSession(){
        return new WorkflowSession(this);
    }

    public boolean isIsInitialized() {
        return isInitialized;
    }
    
    
}

File: dot-net_version/FireWorkflow.Net/FireWorkflow.Net/Engine/RuntimeContext.cs
/**
 * Copyright 2003-2008 非也
 * All rights reserved. 
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation。
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses. *
 * @author 非也,nychen2000@163.com
 * @Revision to .NET 无忧 lwz0721@gmail.com 2010-02
 */
using System;
using System.Collections.Generic;
using System.Text;
using FireWorkflow.Net.Engine.Beanfactory;
using FireWorkflow.Net.Engine.Calendar;
using FireWorkflow.Net.Engine.Definition;
using FireWorkflow.Net.Engine.Impl;
using FireWorkflow.Net.Engine.Persistence;
using FireWorkflow.Net.Engine.Taskinstance;
using FireWorkflow.Net.Kernel;
using FireWorkflow.Net.Engine.Condition;

namespace FireWorkflow.Net.Engine
{
    /// <summary>
    /// RuntimeContext是Fire workflow Engine的总线。所有的服务都挂接在这个总线上，并通过这个总线获取。<br/>
    /// RuntimeContext也是业务代码调用工作流引擎的入口，通过runtimeContext.getWorkflowSession()获得IWorkflowSession 对象，
    /// 然后通过IWorkflowSession调用各种工作流实例对象及其API。<br/>
    /// context管理的各种服务
    /// </summary>
    public class RuntimeContext
    {
        /// <summary>是否已经初始化</summary>
        public Boolean IsInitialized { get; set; }
        //context管理的各种服务

        /// <summary>是否打开流程跟踪，如果打开，则会往T_FF_HIST_TRACE表中插入纪录。</summary>
        public Boolean EnableTrace { get; set; }

        /// <summary>转移条件表达式解析服务</summary>
        private IConditionResolver _conditionResolver = null;
        /// <summary>设置或获取</summary>
        public IConditionResolver ConditionResolver
        {
            get { return _conditionResolver; }
            set
            {
                this._conditionResolver = value;
                if (this._conditionResolver is IRuntimeContextAware)
                {
                    ((IRuntimeContextAware)this._conditionResolver).RuntimeContext = this;
                }
            }
        }

        /// <summary>实例对象存取服务</summary>
        private IPersistenceService _persistenceService = null;
        /// <summary>实例对象存取服务</summary>
        public IPersistenceService PersistenceService
        {
            get { return this._persistenceService; }
            set
            {
                this._persistenceService = value;
                this._persistenceService.RuntimeContext = this;
            }
        }

        /// <summary>流程定义服务，通过该服务获取流程定义</summary>
        private IDefinitionService _definitionService = null;
        /// <summary>流程定义服务，通过该服务获取流程定义</summary>
        public IDefinitionService DefinitionService
        {
            get { return _definitionService; }
            set
            {
                this._definitionService = value;
                this._definitionService.RuntimeContext = this;
            }
        }

        /// <summary>内核管理器</summary>
        private KernelManager _kernelManager = null;
        /// <summary>内核管理器</summary>
        public KernelManager KernelManager
        {
            get { return _kernelManager; }
            set
            {
                this._kernelManager = value;
                //KernelExtensions  Spring.net还没想到方法解决初始化
                this._kernelManager.KernelExtensions.Add("FireWorkflow.Net.Kernel.StartNodeInstance",
                    new List<Kernel.Plugin.IKernelExtension>() { new FireWorkflow.Net.Engine.Kernelextensions.StartNodeInstanceExtension() }
                    );
                this._kernelManager.KernelExtensions.Add("FireWorkflow.Net.Kernel.ActivityInstance",
                    new List<Kernel.Plugin.IKernelExtension>() { new FireWorkflow.Net.Engine.Kernelextensions.ActivityInstanceExtension() }
                    );
                this._kernelManager.KernelExtensions.Add("FireWorkflow.Net.Kernel.SynchronizerInstance",
                    new List<Kernel.Plugin.IKernelExtension>() { new FireWorkflow.Net.Engine.Kernelextensions.SynchronizerInstanceExtension() }
                    );
                this._kernelManager.KernelExtensions.Add("FireWorkflow.Net.Kernel.EndNodeInstance",
                    new List<Kernel.Plugin.IKernelExtension>() { new FireWorkflow.Net.Engine.Kernelextensions.EndNodeInstanceExtension() }
                    );
                this._kernelManager.KernelExtensions.Add("FireWorkflow.Net.Kernel.TransitionInstance",
                    new List<Kernel.Plugin.IKernelExtension>() { new FireWorkflow.Net.Engine.Kernelextensions.TransitionInstanceExtension() }
                    );
                this._kernelManager.KernelExtensions.Add("FireWorkflow.Net.Kernel.LoopInstance",
                    new List<Kernel.Plugin.IKernelExtension>() { new FireWorkflow.Net.Engine.Kernelextensions.LoopInstanceExtension() }
                    );
                this._kernelManager.RuntimeContext = this;
            }
        }

        /// <summary>TaskInstance 管理器，负责TaskInstance的创建、运行、结束。</summary>
        private ITaskInstanceManager _taskInstanceManager = null;
        public ITaskInstanceManager TaskInstanceManager
        {
            get { return _taskInstanceManager; }
            set
            {
                this._taskInstanceManager = value;
                this._taskInstanceManager.RuntimeContext = this;
            }
        }

        /// <summary>日历服务</summary>
        private ICalendarService calendarService = null;

        /// <summary>bean工厂，fire workflow默认使用spring作为其实现</summary>
        private IBeanFactory beanFactory = null;


        //    private ThreadLocal<Object> currentDBSession = new ThreadLocal<Object>();
        //	private ThreadLocal<IFireflowSession> currentFireflowSession = new ThreadLocal<IFireflowSession>();
        //保持一个内部用的beanFactory,类似spring的beanfactory,用于注册Listener,ApplicationHandler,AssignmentHandler的
        //单态实例，可以考虑这个beanfactory是否可以引用spring的beanfactory
        //将Bean的创建工作委派给IBeanFactory,20090228
        //    private Map<String, Object> beanCache = new HashMap<String, Object>();


        public RuntimeContext()
        {
            IsInitialized = false;
            EnableTrace = false;
        }


        /// <summary>
        /// 根据bean的name返回bean的实例。<br/>
        /// Fire workflow RuntimeContext将该工作委派给org.fireflow.engine.beanfactory.IBeanFatory
        /// </summary>
        /// <param name="beanName">Bean Name具体指什么是由IBeanFatory的实现类来决定的。</param>
        /// <returns></returns>
        public Object getBeanByName(String beanName)
        {
            if (beanFactory != null)
            {
                return beanFactory.GetBean(beanName);
            }
            else
            {
                throw new NullReferenceException("The RuntimeContext's beanFactory  can NOT be null");
            }
        }

        //DBSession 放在哪里比较好些？
        //    public void setCurrentDBSession(Object dbSession) {
        //        currentDBSession.set(dbSession);
        //    }
        //
        //    public Object getCurrentDBSession() {
        //        return currentDBSession.get();
        //    }



        

       

        

        public KernelManager getKernelManager()
        {
            return _kernelManager;
        }

        public void setKernelManager(KernelManager arg0)
        {
            this._kernelManager = arg0;
            this._kernelManager.RuntimeContext=this;
        }

        public ICalendarService getCalendarService()
        {
            return calendarService;
        }

        public void setCalendarService(ICalendarService calendarService)
        {
            this.calendarService = calendarService;
            this.calendarService.RuntimeContext=this;
        }

        public IWorkflowSession getWorkflowSession()
        {
            return new WorkflowSession(this);
        }


        public IBeanFactory getBeanFactory()
        {
            return beanFactory;
        }

        public void setBeanFactory(IBeanFactory beanFactory)
        {
            this.beanFactory = beanFactory;
        }

        /// <summary>初始化方法</summary>
        public void initialize()// throws EngineException, KernelException 
        {
            //		System.out.println("执行initialize(),the isInitialized="+this.isInitialized);
            //		System.out.println("看看有没有 受理环节实例"+this.getKenelManager().getWFElementInstance("受理"));
            if (!IsInitialized)
            {
                //			initKenelExtensions();
                initAllNetInstances();
                IsInitialized = true;
            }
        }

        protected void initAllNetInstances()// throws KernelException
        {
            /*
            List<WorkflowProcess> allWfProcess = definitionService.getAllWorkflowProcesses();
            this.getKenelManager().clearAllNetInstance();
            for(int i=0;allWfProcess!=null && i<allWfProcess.Count;i++){
            WorkflowProcess process = allWfProcess.get(i);
            this.getKenelManager().createNetInstance( process);
            }
             */
        }

        public Boolean isEnableTrace()
        {
            return EnableTrace;
        }

        public void setEnableTrace(Boolean enableTrace)
        {
            this.EnableTrace = enableTrace;
        }
    }

}
